test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc

```diff
@@ -52,22 +55,46 @@ name: alternate_protocols_cache
     config_helper_.prependFilter(filter);
 
     upstream_tls_ = true;
+    // This configures the upstream to use the connection grid (automatically
+    // selecting protocol and allowing HTTP/3)
     config_helper_.configureUpstreamTls(/*use_alpn=*/true, /*http3=*/true, alt_cache);
 
     HttpProtocolIntegrationTest::initialize();
   }
 
   void createUpstreams() override {
-    // The integration test infrastructure does not know how to create mixed protocol upstreams.
-    // It will have created an HTTP/3 upstream, so the test needs to create an additional
-    // HTTP/2 upstream manually.
-    ASSERT_EQ(upstreamProtocol(), Http::CodecType::HTTP3);
-    ASSERT_EQ(fake_upstreams_count_, 1);
+    // This function will create 2 upstreams, but Envoy will only point at the
+    // first, the HTTP/2 upstream.
+    skipPortUsageValidation();
     ASSERT_FALSE(autonomous_upstream_);
 
-    auto config = configWithType(Http::CodecType::HTTP2);
-    Network::TransportSocketFactoryPtr factory = createUpstreamTlsContext(config);
-    addFakeUpstream(std::move(factory), Http::CodecType::HTTP2);
+    // Until alt-svc supports different ports, try to get a TCP and UDP fake upstream on the same
+    // port.
+    for (int i = 0; i < 10; ++i) {
+      TRY_ASSERT_MAIN_THREAD {
+        // Make the first upstream HTTP/2
+        auto http2_config = configWithType(Http::CodecType::HTTP2);
+        Network::TransportSocketFactoryPtr http2_factory = createUpstreamTlsContext(http2_config);
+        addFakeUpstream(std::move(http2_factory), Http::CodecType::HTTP2);
+
+        // Make the next upstream is HTTP/3
+        auto http3_config = configWithType(Http::CodecType::HTTP3);
+        Network::TransportSocketFactoryPtr http3_factory = createUpstreamTlsContext(http3_config);
+        // If the UDP port is in use, this will throw an exception and get caught below.
+        fake_upstreams_.emplace_back(std::make_unique<FakeUpstream>(
+            std::move(http3_factory), fake_upstreams_[0]->localAddress()->ip()->port(), version_,
+            http3_config));
+        addFakeUpstream(std::move(http3_factory), Http::CodecType::HTTP3);
```

Comment:
I think this is a use-after-move since http3_factory is moved on line 85?

Reply:
yeah and spurious - we don't want 3 upstreams.  I'd started with that one, added manual creation for the specific port, and forgot to remove.  fixed!